Wrappers/Python/test/test_Gradient.py

```diff
@@ -17,555 +17,382 @@
-        numpy.testing.assert_array_equal(out_direct.get_item(2).as_array(), gold_direct.get_item(2).as_array())
-        numpy.testing.assert_array_equal(out_direct.get_item(3).as_array(), gold_direct.get_item(3).as_array())
-
-        #print("GradientOperator, 4D, bnd_cond='Neumann', adjoint")
-        numpy.testing.assert_array_equal(out_adjoint.as_array(), gold_adjoint.as_array())
+        print( " Test Failed ")
+        print( " ImageGeometry {} \n".format(geom))
+        print( " Bnd Cond {} ".format(bnd))                                
+        print( " Backend {} ".format(backend))
+        print( " Method {} ".format(method))
+        print( " Correlation {} ".format(corr)) 
+        if split is not None:
+            print( " Split {} ".format(split))        
+            
 
     def test_GradientOperator_linearity(self):
 
-        nc, nz, ny, nx = 3, 4, 5, 6
-        ig = ImageGeometry(voxel_num_x=nx, voxel_num_y=ny, voxel_num_z=nz, channels=nc)
-
-        grad = GradientOperator(ig, bnd_cond='Neumann', correlation='SpaceChannels', backend='c')   
-        self.assertTrue(LinearOperator.dot_test(grad))
-
-        grad = GradientOperator(ig, bnd_cond='Periodic', correlation='SpaceChannels', backend='c')
-        self.assertTrue(LinearOperator.dot_test(grad))
-
-        grad = GradientOperator(ig, bnd_cond='Neumann', correlation='SpaceChannels', backend='numpy')
-        self.assertTrue(LinearOperator.dot_test(grad))
-
-        grad = GradientOperator(ig, bnd_cond='Periodic', correlation='SpaceChannels', backend='numpy')
-        self.assertTrue(LinearOperator.dot_test(grad))
-        
-    def test_Gradient_c_numpy_voxel(self):
-        
-        numpy.random.seed(5)
-        
-        print("Test GradientOperator for 2D Geometry, ")
-        
-        ny, nx, nz = 3, 4, 5
-        ig = ImageGeometry(voxel_num_y = ny, voxel_num_x = nx, voxel_size_x=0.1, voxel_size_y=0.5) 
+        for geom in self.list_geometries:            
+            for bnd in self.bconditions:
+                for backend in self.backend:
+                    for corr in self.correlation:
+                        for method in self.method:                       
+                                                                                          
+                            Grad = GradientOperator(geom, 
+                                                    bnd_cond = bnd,
+                                                    backend = backend, 
+                                                    correlation = corr, method = method)
+                            try:                                                    
+                                self.assertTrue(LinearOperator.dot_test(Grad, seed=10))
+                            except:    
+                                self.print_assertion_info(geom,bnd,backend,corr,method,None)
+                                raise
```

Comment:
Why is this required? Wouldn't it be possible to get the assertion error without `print_assertion_info`?

At any rate, I understand why you want to use the nested `for` loops to run all these tests that perform the `dot_test` for many configurations. But how it's written now, the tests will only run until one fails. 

Say for instance 2 tests fail, we will only get reported one. Wouldn't it be better to run all the tests and compile a list of results and fail at the end reporting all the tests that failed?

Reply:
Looking backward, I believe I could live with a nested loop and failure as is now, though it's not great.

Reply:
yes, that's useful to discern which test has failed